pool.js

```diff
@@ -0,0 +1,53 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+
+function fastifyMysql (fastify, options, next) {
+  const name = options.name
+  delete options.name
+
+  const usePromise = options.promise
+  delete options.promise
+
+  const mysql = usePromise ? require('mysql2/promise') : require('mysql2')
+
+  const pool = mysql.createPool(options.connectionString || options)
+  const db = {
+    connect: onConnect => pool.getConnection(onConnect),
+    pool: pool,
+    query: pool.query.bind(pool),
+    end: pool.end.bind(pool),
+    getConnection: pool.getConnection.bind(pool)
+  }
+
+  if (name) {
+    if (!fastify.mysql) {
+      fastify.decorate('mysql', {})
+    }
+    fastify.mysql[name] = db
+  } else {
+    if (fastify.mysql) {
+      next(new Error('fastify-mysql has already registered'))
+    } else {
+      fastify.mysql = db
+    }
+  }
+
+  if (usePromise) {
+    fastify.addHook('onClose', async (fastify, done) => {
+      fastify.mysql = null
+      await pool.end()
```

Comment:
You can't use `async` here. You will have to do `return pool.end()`. This needs to run on Node6, which does not have async/await yet.